### ISSUE TITLE
Matching of Licenses During Analysis

### ISSUE THREAD

**Issue by msymons** [DependencyTrack/dependency-track#248]
Dependency-Track v3.3.1 is parsing License information from both CycloneDX BOM and from Dependency-Check XML and, when licensing info is present, it is mostly not then linking the license to a known License.

ie, instead of "Apache-2.0" (linking to ` ../license/?licenseId=Apache-2.0`), there is simple text displayed:

 - The Apache Software License, Version 2.0
 - Apache License 2.0
 - Apache Software Licenses

I am not sure if this would be an enhancement or a defect, but my expectation is that all 3 of the above examples would have been identified as "Apache-2.0" (complete with link).

Using the last example (Purl = `pkg:maven/org.slf4j/log4j-over-slf4j@1.7.25?type=jar`):

```
<licenses>
  <license>
    <name>Apache Software Licenses</name>
    <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
  </license>
</licenses>
```
..there seems to be enough info (including the url) to match the license correctly.

The components that have licenses that *are* linked all seem to have a filename.  Example:

`glassfish-embedded-all-3.1.2.2.jar (shaded: org.jvnet.libpam4j:libpam4j:1.3)`
Purl:  `pkg:maven/org.jvnet.libpam4j/libpam4j@1.3`

Dependency-Track lists this as CDDL 1.0 - although libpam4j itself has an MIT license.

- Tested with BOM using cyclonedx-maven-plugin-1.2.0
- Tested with XML using dependency-check-jenkins-plugin-4.0.0

**Comment by stevespringett** [comment]
In the case of Maven, it doesn't standardize on SPDX, whereas NPM for example, does. When generating a CycloneDX BOM from NPM, most of the modules (except for internal or grandfathered) will contain valid SPDX license IDs. In Maven, the license is just plain text and the specification doesn't provide any guidance on what should be written there.

Ideally, the license matching logic would be in each of the CycloneDX implementations so that they could generate more accurate license info in the BOM. So instead of a license name field in the bom, you'd have a license id field instead. This is already supported, but the fuzzy logic to do the matching in the Maven plugin isn't there yet. Refer to: https://github.com/CycloneDX/cyclonedx-maven-plugin/blob/master/src/main/java/org/cyclonedx/maven/BaseCycloneDxMojo.java#L232

On the Dependency-Track side, we also need to account for SPDX expressions. This relates to #170. So it's not quite as simple as doing fuzzy matching. A bit more involved there.

**Comment by stevespringett** [comment]
Added enhancement: https://github.com/CycloneDX/cyclonedx-core-java/issues/1

**Comment by msymons** [comment]
I have not yet analysed an NPM project - but was about to start doing so.  I look forward to having a slightly easier time there!

Getting back to Maven, I see that in my initial set of projects (14 CycloneDX BOM upload and 1 XML from Dependency-Check Plugin) I have some 80 "jackson" components.  eg

`pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.8.7?type=jar`

..and not one has license information contained in the component POM.  The license info is in the parent POM.   In the example above,  this would be [jackson-parent-2.8.pom](http://central.maven.org/maven2/com/fasterxml/jackson/jackson-parent/2.8/jackson-parent-2.8.pom)

In the case where the component POM contains *no* license information, could the cyclonedx-maven-plugin extract the license information from the parent POM and include it in the BoM element for each relevent component?  If so, I'll log an enhancement issue against cyclonedx-maven-plugin. If valid, I see this as an enhancement that is orthogonal to the implementation of fuzzy matching  (but they would complement each other).

I gave the example for jackson components, but had the exact same problem for dropwizard (where I had some 50 components).    ie, the ability of the plugin to extract the license info from the parent POM could save a *lot* of manual checking and fixing within Dependency-Track.

**Comment by stevespringett** [comment]
Yes, Jacksons a very good example. If that's the way the license info is inherited, then this would be a good enhancement. If you have a sample project on github that would be ideal, otherwise I can try to create one.

**Comment by msymons** [comment]
The only projects I have are private repos.

I would think that [dropwizard](https://github.com/dropwizard/dropwizard) would be good for a test-case.  Lots of modules that are all parented by dropwizard-parent (which has the license info), with a comprehensive [getting started](https://www.dropwizard.io/1.3.5/docs/getting-started.html) guide.

**Comment by stevespringett** [comment]
The parent pom isn’t a Maven artifact from their perspective when resolving from a project. So it’s easily missed. This is why the license (and other elements) were missing. I did manage to come up with a way in the CycloneDX Maven plugin to resolve it anyway and do so in a way that would account for multiple levels of parent/child relationships. I’ll hopefully have something checked in in a day or two for you to test. Using dropwizard for testing.

**Comment by stevespringett** [comment]
I've published CycloneDX code that results in much improved metadata extraction. This takes into consideration the entire inheritance tree of every single direct and transitive dependency.

I've also corrected issues that resulted in duplicate dependencies being introduced in the BOM. The duplicates would have never shown up in Dependency-Track, but the BOM itself had them. They should be gone now.

These changes are available in CycloneDX Maven Plugin v1.3.0

**Comment by msymons** [comment]
I have used CycloneDX Maven Plugin v1.3.0 in multiple projects and compared the BOM output with that created by v1.2.0.  I have not yet seen one project where the "hit-rate" has not significantly improved.

One project improved the license count from 18 to 107!

So, thanks muchly for the speedy response.
